VNC не корректно создает URL для подключения к ВМ

tihon49 opened the pull request:
# Описание изменений

Реализован механизм управления VNC-сессиями для виртуальных машин в Open vAIR. Sub-модуль обеспечивает потокобезопасное выделение портов, автоматический запуск `websockify` и корректное завершение VNC-сессий.

---

# 🚀 Ключевые изменения:

-  **Полностью переписан метод vnc() в LibvirtDriver:**
    - cтарй код запускал websockify напрямую с жестко заданными портами 
    - новый код использует VNCManager для динамического выделения портов
-  **Добавлен новый VNCManager модуль (/vnc/):**
    - singleton паттерн для управления портами
        - без singleton: каждый процесс создавал бы свой VNCManager что вызывало бы конфликты портов!
    - thread-safe операции 
    - автоматическая очистка мертвых сессий
    - умное выделение портов в диапазоне 6100-6999
    - `exceptions.py` — набор специализированных исключений для управления VNC.
    - `__init__.py` — экспорт ключевых компонентов модуля.
    - Запуск `websockify` в режиме демона с привязкой к конкретной VM.
    - Поиск PID процесса по порту (`lsof`) и его завершение при остановке VM.
    - Удаление сессии и освобождение порта после завершения.

- **Конфигурация:**
  - В `config.py` добавлены параметры:
    - `VNC_WS_PORT_START`, `VNC_WS_PORT_END` — диапазон WebSocket-портов.
    - `VNC_LOCK_FILE`, `VNC_STATE_FILE`, `VNC_MAX_SESSIONS` — зарезервированы для возможного расширения (не используются сейчас).
  
---

# Требования к тестированию

1. **Функциональное тестирование VNC:**
   - Запустить виртуальную машину через API.
   - Вызвать метод `vnc()` и получить URL для подключения.
   - Проверить успешное открытие VNC-сессии в браузере через noVNC.
   - Остановить VM и убедиться, что VNC-сессия завершена (`websockify` процесс завершён, порт освобождён).

2. **Проверка многопоточности и конфликтов:**
   - Параллельно запустить несколько VM и VNC-сессий.
   - Убедиться, что порты не конфликтуют и корректно освобождаются при остановке VM.

3. **Отказоустойчивость:**
   - Симулировать сбой запуска `websockify` (например, занять порт вручную) и убедиться в выбрасывании `VncSessionStartupError`.
   - Проверить, что исключения логируются и не блокируют дальнейшую работу.

---

# Критерии приёмки

- VNC-сессии успешно запускаются, завершаются и не оставляют висячих процессов.
- В случае ошибки предоставляется информативный лог и корректное исключение.
- Использование WebSocket-портов остаётся в пределах заданного диапазона.
- Нет утечек ресурсов (процессы `websockify`, блокировки, порты).

---

# Заключение

Данный PR упрощает и стабилизирует жизненный цикл VNC-сессий, улучшая масштабируемость и поддержку в кластере виртуальных машин. В будущем модуль можно расширить для работы с распределённым состоянием и более сложными политиками порт-менеджмента.

tihon49 committed:
feat: Implement scalable VNC port management system

- Add VncPortPoolManager for thread-safe port allocation (6100-6999 range)
- Add WebsockifyProcessManager for websockify lifecycle management
- Add VncSessionCoordinator for atomic resource coordination
- Refactor LibvirtDriver.vnc() to use new port management system
- Add automatic VNC resource cleanup on VM shutdown
- Add VNC cleanup daemon script for stale resource management
- Add comprehensive error handling with custom exceptions
- Add thread-safe file-based state persistence with fcntl locking
- Support concurrent startup of 900+ VMs without port conflicts
- Resolve 'Address already in use' error for mass VM deployments

This fixes the critical issue where multiple VMs starting simultaneously
would fail due to WebSocket port conflicts, enabling reliable scaling
to hundreds of concurrent VNC sessions.

tihon49 committed:
refactor: Move VNC cleanup daemon and replace os calls with execute

- Move VncCleanupDaemon from scripts/ to virtual_machines/vnc/ module
- Remove scripts/ directory entirely - logic now in proper module location
- Replace all direct os.* calls with openvair.libs.cli.executor.execute()
- Update port_pool_manager.py: use execute() for mv, chmod, kill operations
- Update process_manager.py: use execute() for kill, lsof, pgrep operations
- Maintain consistent API usage throughout VNC management system
- Remove temporary test files

This improves code organization by keeping business logic in modules
rather than utility scripts, and ensures consistent use of internal
execute() API for all system operations.

tihon49 committed:
refactor: simplify VNC system by removing unused methods and timeouts

- Remove unused methods: cleanup_stale_resources, get_system_status, list_active_sessions, get_session_info, stop_by_port, get_process_count
- Replace mark_port_in_use with simpler update_port_pid method
- Remove websockify timeout parameter (--timeout 300)
- Delete cleanup_daemon.py file
- Reduce codebase from 1057 to 940 lines (-11%)
- Maintain core functionality: thread-safe port allocation, atomic operations, proper cleanup

tihon49 committed:
_

tihon49 committed:
renamed vnc manager class

tihon49 committed:
refacroring vnc manager

tihon49 committed:
refactored

tihon49 committed:
Merge branch 'dev' into bug/VNC_already_in_use

tihon49 committed:
refacroted: simplify VncManager by removing overengineering

tihon49 committed:
renamed: VNCManager

tihon49 committed:
refactored: vm_name variable

tihon49 committed:
fix: noqa in exception

tihon49 committed:
hotfix: vm_id -> vm_name

tihon49 committed:
hotfix: description

tihon49 committed:
hotfix: description

tihon49 committed:
Merge branch 'dev' into bug/VNC_already_in_use

VladimirTrue commented on openvair/modules/virtual_machines/config.py:
```diff
@@ -10,4 +10,11 @@
+VNC_WS_PORT_END = 6999  # 900 ports available
+VNC_LOCK_FILE = '/var/lock/openvair_vnc_ports.lock'
+VNC_STATE_FILE = '/opt/aero/openvair/data/vnc_ports.json'
+VNC_MAX_SESSIONS = 800  # Leave some buffer
```
Не думаю, что эти коменты нужны

tihon49 replied:
fixed

VladimirTrue commented on openvair/modules/virtual_machines/domain/libvirt2/driver.py:
```diff
@@ -179,23 +181,59 @@ def turn_off(self) -> Dict:
+        """
+        vnc_session = self.vm_info.get('vnc_session')
+        if not vnc_session:
+            return  # No VNC session to cleanup
```
тоже комменты не оч нужные мне кажутся. По всему pr далее отмечать это не буду=)

tihon49 replied:
fixed

VladimirTrue reviewed (dismissed):
Столкнулся с исходной проблемой already in use.
**UPD:** некорректно протестил.

Так же имеет смысл доработать идёю с реализацией синглтона, если вообще не использовать его как паттерн. Мне кажется не лучшим решением, это перестанет нормально работать в тот момент, когда будет добавлена поддержка нескольких менеджеров для rpc-сервера, которые смогут работать параллельно и каждый, создавая своё эксземпляр. Возможно поискать либы для управления vnc сессиями.

tihon49 committed:
deleted: global vnc_manager variable

tihon49 committed:
refactored: del functions. Use VNCManager instead

tihon49 committed:
update: unnecessary comments removed

tihon49 committed:
update: unnecessary comments removed

tihon49 committed:
update: coma added

tihon49 commented on openvair/modules/virtual_machines/domain/libvirt2/driver.py:
```diff
@@ -213,36 +251,43 @@ def vnc(self) -> Dict:
-                ),
+            # Use the VNC manager for port allocation and websockify
+            session_result = start_vnc_session(
+                vm_name=vm_name, vnc_host='localhost', vnc_port=vnc_port
```
fixed

tihon49 commented on openvair/modules/virtual_machines/vnc/__init__.py:
```diff
@@ -0,0 +1,34 @@
+    'VncPortAllocationError',
+    'VncProcessNotFoundError',
+    'VncSessionStartupError',
+]
```
Как вариант. Можно убрать.

tihon49 commented on openvair/modules/virtual_machines/vnc/manager.py:
```diff
@@ -0,0 +1,579 @@
+
+
+# Global singleton instance
+vnc_manager = VNCManager()
```
Переменная и функции: fixed.
Singletone: в процессе обдумывания.

tihon49 committed:
ruff format and mypy check passed

tihon49 committed:
update: unnecessary comments removed

tihon49 committed:
refactoring: the general logic is moved to libs.

tihon49 committed:
ruff and mypy check padded

tihon49 committed:
noqa comments added

tihon49 committed:
fixed: imports

tihon49 committed:
added: module docstring

tihon49 committed:
added: exceptions docstring

dr-wh0 reviewed (approved):
Протестировал, всё отлично работает

tihon49 committed:
Refactor process termination using psutil

Replace manual 'kill' command with kill_process_by_pid utility using psutil for cleaner and more reliable process termination

tihon49 committed:
Added: use VncSessionStartupError exception

tihon49 committed:
Improve websockify process restoration logic

Refactor VNCManager to handle missing cmdline in process info

tihon49 commented:
> Столкнулся с исходной проблемой already in use. **UPD:** некорректно протестил.
> 
> Так же имеет смысл доработать идёю с реализацией синглтона, если вообще не использовать его как паттерн. Мне кажется не лучшим решением, это перестанет нормально работать в тот момент, когда будет добавлена поддержка нескольких менеджеров для rpc-сервера, которые смогут работать параллельно и каждый, создавая своё эксземпляр. Возможно поискать либы для управления vnc сессиями.

В данный момент singleton позволяет решить проблемы:
- race conditions (несколько экземпляров могут выделить один порт) 
- рассинхронизация состояния - разные экземпляры будут иметь разное представление о занятых портах

Если запустить 10 ВМ и для каждй из них открыть VNC-сессию, то будет 10 экземпляров класса `VNCManager`. 
С одной стороны у нас есть проверка на "занятость" порта в методе `is_port_free`, но это не защитит от race condition: первый процесс проверят порт на "занятость" и еще один процесс проверят этот же порт на "занятость". После чего один из процессов успевает занять порт, но и второй процесс тоже назначает тот же порт. В результате получаем ту же проблему, которую пытались решить данным PR: некорректная работа выделения свободных портов.

Возможно сделать через `DI`, но мне кажется что это избыточно для данной задачи.
